proposals/lip-0044.md

```diff
@@ -190,234 +216,320 @@ genesisDataSchema = {
+It returns the data structure stored in the genesis data substore.
 
 #### setValidatorBLSKey
 
 This function sets the BLS key of a validator account. This update is only possible if the current BLS key of the validator is equal to a predefined constant invalid BLS key.
-
-##### Parameters
-
-This function has the following input parameters in the order given below:
-
-* `validatorAddress`: A 20 bytes value identifying the validator account in the validators data substore.
-* `proofOfPossession`: The [proof of possession](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0038.md#public-key-registration-and-proof-of-possession) for `blsKey`.
-* `blsKey`: A 48 bytes value, specifying the new BLS key of the validator.
-
-##### Returns
-
-* `success`: A boolean, indicating success or failure of the function execution.
-
-##### Execution
-
-This function checks that there exists a validator account registered for the input `validatorAddress`, that the input BLS key has not been registered before in the chain, that the BLS key of the validator account is equal to a certain constant invalid BLS key, and that the proof of possession for the BLS key is valid. Finally, it updates the BLS key of the validator account.
+It checks that there exists a validator account registered for the input `validatorAddress`, that the input BLS key has not been registered before in the chain, that the BLS key of the validator account is equal to a certain constant invalid BLS key, and that the input proof of possession `proofOfPossession` for the BLS key is valid. Finally, it updates the BLS key of the validator account to the input `blsKey`.
 
 ```python
-setValidatorBLSKey(validatorAddress, proofOfPossession, blsKey):
+def setValidatorBLSKey(validatorAddress: Address, proofOfPossession: ProofOfPossession, blsKey: BLSKey) -> None:
     if no entry in the validators data substore exist with storeKey == validatorAddress:
-        return False
+        emitPersistentEvent(
+            moduleID = MODULE_ID_VALIDATORS,
+            typeID = TYPE_ID_BLS_KEY_REGISTRATION,
+            data = {
+                "blsKey": blsKey,
+                "result": KEY_REG_RESULT_NO_VALIDATOR
+            },
+            topics=[validatorAddress]
+        )
+        raise Exception('This address is registered as validator. Only validators can register a BLS key.')
     if there exists an entry in the registered BLS keys substore with storeKey == blsKey:
-        return False
+        emitPersistentEvent(
+            moduleID = MODULE_ID_VALIDATORS,
+            typeID = TYPE_ID_BLS_KEY_REGISTRATION,
+            data = {
+                "blsKey": blsKey,
+                "result": KEY_REG_RESULT_DUPLICATE_BLS_KEY
+            },
+            topics=[validatorAddress]
+        )
+        raise Exception('This address is already registered as validator.')
     if validatorAccount(validatorAddress).blsKey != INVALID_BLS_KEY:
```

Comment:
Note that this check is already done in [LP 50](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0050.md#verification-1) which is the only place this API is called. Would it make sense to allow BLS key updates in general for the future although write now we do not use this feature? At the moment it is odd that we allow generator key updates (although we do no use them as the would break the PoM), but forbid BLS key updates. As far as the Validator module is concerned, the BLS keys can be updated and nothing breaks here.

Reply:
after discussion, we remove the check